lib/PrintAsClang/ModuleContentsWriter.cpp

```diff
@@ -780,44 +804,65 @@ class ModuleWriter {
           std::mismatch(lhsProtos.begin(), lhsProtos.end(), rhsProtos.begin(),
                         [] (const ProtocolDecl *nextLHSProto,
                             const ProtocolDecl *nextRHSProto) {
-          return nextLHSProto->getName() != nextRHSProto->getName();
+          return nextLHSProto->getName() == nextRHSProto->getName();
         });
         if (mismatch.first != lhsProtos.end()) {
           StringRef lhsProtoName = (*mismatch.first)->getName().str();
-          return lhsProtoName.compare((*mismatch.second)->getName().str());
+          result = lhsProtoName.compare((*mismatch.second)->getName().str());
+          if (result != 0)
+            return result;
         }
       }
 
-      // Still nothing? Fine, we'll pick the one with the alphabetically first
-      // member instead.
+      // Still nothing? Fine, we'll look for a difference between the members.
       {
-        auto mismatch = std::mismatch(
-            cast<ExtensionDecl>(*lhs)->getAllMembers().begin(),
-            cast<ExtensionDecl>(*lhs)->getAllMembers().end(),
-            cast<ExtensionDecl>(*rhs)->getAllMembers().begin(),
-            [](const Decl *nextLHSDecl, const Decl *nextRHSDecl) {
-              if (isa<ValueDecl>(nextLHSDecl) && isa<ValueDecl>(nextRHSDecl)) {
-                return cast<ValueDecl>(nextLHSDecl)->getName() !=
-                       cast<ValueDecl>(nextRHSDecl)->getName();
-              }
-              return isa<ValueDecl>(nextLHSDecl) != isa<ValueDecl>(nextRHSDecl);
-            });
-        if (mismatch.first !=
-            cast<ExtensionDecl>(*lhs)->getAllMembers().end()) {
-          auto *lhsMember = dyn_cast<ValueDecl>(*mismatch.first),
-               *rhsMember = dyn_cast<ValueDecl>(*mismatch.second);
+        for (auto pair : llvm::zip_equal(lhsMembers, rhsMembers)) {
```

Comment:
This function starts to be quite big.  I wonder if we should factor out some of these comparisons into separate helper functions.

Reply:
Pushed a pretty comprehensive refactoring just now.